src/test/unit_tests/test_s3select.js

```diff
@@ -0,0 +1,205 @@
+        const s3select = new s3select_utils.S3SelectStream(select_args);
+        s3select.destroy();
+    });
+
+    mocha.it('select star', async function() {
+        const select_args = {
+            query: "select * from stdin;",
+            input_format: "CSV",
+            input_serialization_format: {"FieldDelimiter": ",", "RecordDelimiter": "\n"},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const output = await run_sql(select_args, csv_small_str);
+        assert.strictEqual(output, csv_small_str);
+    });
+
+    mocha.it('select sum with where', async function() {
+        const select_args = {
+            //sum every third number in column 2
+            query: "select sum(int(_2)) from stdin where int(_1) % 3 = 1;",
+            input_format: "CSV",
+            input_serialization_format: {"FieldDelimiter": ",", "RecordDelimiter": "\n"},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const output = await run_sql(select_args, csv_small_str);
+        assert.strictEqual(output, "1554");
+    });
+
+
+    mocha.it('select with header name', async function() {
+        const select_args = {
+            //select column named 'c' where int value of column 'a' is less than 5
+            query: "select c from stdin where int(a) < 5;",
+            input_format: "CSV",
+            input_serialization_format: {"FieldDelimiter": ",", "RecordDelimiter": "\n", "FileHeaderInfo": "USE"},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const output = await run_sql(select_args, csv_small_str);
+        assert.strictEqual(output, "339\n741\n53\n233\n55\n");
+    });
+
+    mocha.it('select json', async function() {
+        const select_args = {
+            //select intf field from the jsons in the array (S3Object[*])
+            query: "select _1.intf from S3Object[*] s;",
+            input_format: "JSON",
+            input_serialization_format: {},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const output = await run_sql(select_args, json_str);
+        assert.strictEqual(output, "3\n123\n");
+    });
+
+    mocha.it('double pipe close', async function() {
+        const readable = stream.Readable.from(["this is an input string."]);
+        const collect = new CollectChunks();
+        const transform = new DummyTransform(collect);
+        stream_utils.pipeline([transform, collect], true /*res is a write stream, no need for resume*/);
+        await stream_utils.pipeline([readable, transform], true);
+        const output = Buffer.concat(collect.chunks).toString();
+        assert.strictEqual(output, "this is an input string.");
+        assert(readable.emit('error', new Error('readale error')));
+        assert(collect.closed);
+    });
+
+    mocha.it('async loop error', async function() {
```

Comment:
Where is the error? I don't understand the test

Reply:
In simulate_async_loop() (last line).